blockchain/indexers/addrindex.go

```diff
@@ -0,0 +1,932 @@
+// Copyright (c) 2016 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package indexers
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/btcsuite/btcd/blockchain"
+	"github.com/btcsuite/btcd/chaincfg"
+	database "github.com/btcsuite/btcd/database2"
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+)
+
+const (
+	// addrIndexName is the human-readable name for the index.
+	addrIndexName = "address index"
+
+	// level0MaxEntries is the maximum number of transactions that are
+	// stored in level 0 of an address index entry.  Subsequent levels store
+	// 2^n * level0MaxEntries entries, or in words, double the maximum of
+	// the previous level.
+	level0MaxEntries = 8
+
+	// addrKeySize is the number of bytes an address key consumes in the
+	// index.  It consists of 1 byte address type + 20 bytes hash160.
+	addrKeySize = 1 + 20
```

Comment:
As part of the new segwit softfork, two new address types have been introduced. One of the new address types, pay-to-witness-script-hash or P2WSH, has the following "standard" pkScript template: 
  * `0 <32-byte-hash>`

Functionally, this is just regular p2sh, but now with a 32-byte hash (sha256) rather than a 20-byte hash (hash160).

As a result, all "traditional" addresses can no longer be uniquely pattern-matched with a 20-byte data push. I think it'd be prudent to modify the way the format of the `addrKey`, in addition to the way it's calculated once a "standard" address type is encountered. 

I propose the `addrKeySize` be promoted to be 33 bytes (1 byte address type + 32 byte sha256). While indexing a transaction's inputs, and outputs, whenever a known address is encountered if it contains a salient 20-byte data push (`btcutil.Address.ScriptAddress()`), then we hash that to create a 32-byte key. If the data push is instead 32-bytes, we just use that directly, skipping the intermediate hashing phase. 

Thoughts?